File: docs/tests/test_howto.cpp
#include <gtest/gtest.h>
#include <opendaq/opendaq.h>
#include <fstream>

#include <thread>
#include "docs_test_helpers.h"

using HowToTest = testing::Test;

BEGIN_NAMESPACE_OPENDAQ

// Corresponding document: Antora/modules/howto_guides/pages/howto_connect_to_device.adoc
TEST_F(HowToTest, ConnectToDevice1)
{
    SKIP_TEST_MAC_CI;

    InstancePtr server = docs_test_helpers::setupSimulatorServers();
    daq::InstancePtr instance = daq::Instance();

    daq::DevicePtr device = instance.addDevice("daq.opcua://127.0.0.1/");
    ASSERT_TRUE(device.assigned());
}

// Corresponding document: Antora/modules/howto_guides/pages/howto_connect_to_device.adoc
TEST_F(HowToTest, ConnectToDevice2)
{
    daq::InstancePtr instance = daq::Instance();

    ListPtr<IDevice> devices = List<IDevice>();
    for (const auto& deviceInfo : instance.getAvailableDevices())
        if (deviceInfo.getConnectionString().toStdString().find("daqref://") != std::string::npos)
            devices.pushBack(instance.addDevice(deviceInfo.getConnectionString()));

    ASSERT_EQ(devices.getCount(), (SizeT) 2);
}

void printProperty(const PropertyPtr& /*prop*/)
{
    //    std::cout << prop.getName() << std::endl;
}

// Corresponding document: Antora/modules/howto_guides/pages/howto_add_function_block.adoc
TEST_F(HowToTest, AddFunctionBlock)
{
    // Create an openDAQ(TM) Instance, loading modules from the current directory
    InstancePtr instance = Instance();

    // Add simulated device
    DevicePtr device = instance.addDevice("daqref://device0");

    // Get available Function Block types
    DictPtr<IString, IFunctionBlockType> functionBlockTypes = instance.getAvailableFunctionBlockTypes();
    for (const auto& functionBlockType : functionBlockTypes.getKeys())
        std::cout << functionBlockType << std::endl;

    // If there is no Statistics Function Block available, exit with an error
    if (!functionBlockTypes.hasKey("ref_fb_module_statistics"))
        ASSERT_TRUE(false) << "Function Block does not exist";

    // Add Function Block on the host computer
    FunctionBlockPtr functionBlock = instance.addFunctionBlock("ref_fb_module_statistics");

    // Print Function Block type info
    FunctionBlockTypePtr functionBlockType = functionBlock.getFunctionBlockType();
    std::cout << functionBlockType.getId() << std::endl;
    std::cout << functionBlockType.getName() << std::endl;
    std::cout << functionBlockType.getDescription() << std::endl;
}

// Corresponding document: Antora/modules/howto_guides/pages/howto_configure_function_block.adoc
TEST_F(HowToTest, ConfigureFunctionBlock)
{
    // Create an openDAQ(TM) Instance, loading modules from the current directory
    InstancePtr instance = Instance();

    // Add simulated device
    DevicePtr device = instance.addDevice("daqref://device0");

    // Add Function Block on the host computer
    FunctionBlockPtr functionBlock = instance.addFunctionBlock("ref_fb_module_statistics");

    // List properties of the Function Block
    ListPtr<IProperty> functionBlockProperties = functionBlock.getVisibleProperties();
    for (const auto& prop : functionBlockProperties)
        std::cout << prop.getName() << std::endl;

    // Print current block size
    Int currentBlockSize = functionBlock.getPropertyValue("BlockSize");
    std::cout << "Current block size is " << currentBlockSize << std::endl;
    // Configure the properties of the Function Block
    functionBlock.setPropertyValue("BlockSize", 100);

    // Connect the first Signal of the first Channel from the Device to the first Input Port on the Function Block
    functionBlock.getInputPorts()[0].connect(device.getChannels()[0].getSignals()[0]);
    // Read data from the Signal
    // ...

    // Get the output Signal of the Function Block
    SignalPtr outputSignal = functionBlock.getSignals()[0];

    std::cout << outputSignal.getDescriptor().getName() << std::endl;
}

// Corresponding document: Antora/modules/howto_guides/pages/howto_save_load_configuration.adoc
TEST_F(HowToTest, SaveLoadConfiguration)
{
    {
        InstancePtr instance = Instance();

        // Save Configuration to string
        std::string jsonStr = instance.saveConfiguration();
        // Write Configuration string to file
        std::ofstream configFile("config.json");
        configFile << jsonStr;
        configFile.close();
    }

    {
        InstancePtr instance = Instance();

        // Read Configuration from file
        std::ifstream configFile("config.json");
        std::stringstream jsonStr;
        jsonStr << configFile.rdbuf();
        // Load Configuration from string
        instance.loadConfiguration(jsonStr.str());
    }
}

TEST_F(HowToTest, InstanceConfiguration)
{
    InstanceBuilderPtr builder =
        InstanceBuilder().setGlobalLogLevel(LogLevel::Info).setModulePath("").setSchedulerWorkerNum(1).setRootDevice("daqref://device0");

    InstancePtr instance = builder.build();
    daq::DevicePtr device = instance.getRootDevice();
    ASSERT_TRUE(device.assigned());
}

END_NAMESPACE_OPENDAQ

File: bindings/python/tests/test_how_to.py
#!/usr/bin/env python

import unittest
import opendaq_test
import opendaq

class TestDocumentationHowTo(opendaq_test.TestCase):
    # Corresponding document: Antora/modules/howto_guides/pages/howto_add_function_block.adoc
    def test_add_function_block(self):
        # Create an openDAQ(TM) Instance, loading modules from the current directory
        instance = opendaq.Instance()

        # Add simulated device
        device = instance.add_device('daqref://device0')
        
        # Get available Function Block types
        function_block_types = instance.available_function_block_types
        for function_block_type in function_block_types.keys():
            print(function_block_type)

        # If there is no Statistics Function Block available, exit with an error
        if not "RefFBModuleStatistics" in function_block_types.keys():
            self.assertTrue(False, "Function block not found")

        # Add Function Block on the host computer
        function_block = instance.add_function_block("RefFBModuleStatistics")

        # Print Function Block type info
        function_block_type = function_block.function_block_type
        print(function_block_type.id)
        print(function_block_type.name)
        print(function_block_type.description)

    # Corresponding document: Antora/modules/howto_guides/pages/howto_configure_function_block.adoc
    def test_configure_function_block(self):
        # Create an openDAQ(TM) Instance, loading modules from the current directory
        instance = opendaq.Instance()

        # Add simulated device
        device = instance.add_device('daqref://device0')

        # Add Function Block on the host computer
        function_block = instance.add_function_block("RefFBModuleStatistics")

        # List properties of the Function Block
        function_block_properties = function_block.visible_properties
        for prop in function_block_properties:
            print(prop.name)

        # Print current block size
        currentBlockSize = function_block.get_property_value("BlockSize")
        print("Current block size is %d" % (currentBlockSize))
        # Configure the properties of the Function Block
        function_block.set_property_value("BlockSize", 100)

        # Connect the first Signal of the first Channel from the Device to the first Input Port on the Function Block
        function_block.input_ports[0].connect(device.channels[0].signals[0])
        # Read data from the first Signal of the Function Block
        # ...
    
        # Get the output Signal of the Function Block
        output_signal = function_block.signals[0]

        print(output_signal.name)
        
    # Corresponding document: Antora/modules/howto_guides/pages/howto_save_load_configuration.adoc
    def test_save_load_configuration(self):
        instance = opendaq.Instance()

        # Save Configuration to string
        json_str = instance.save_configuration()
        # Write Configuration string to file
        config_file = open("config.json", "w")
        config_file.write(json_str)
        config_file.close()

        instance = opendaq.Instance()

        # Read Configuration from file
        config_file = open("config.json", "r")
        json_str = config_file.read()
        config_file.close()
        # Load Configuration from string
        instance.load_configuration(json_str)

if __name__ == '__main__':
    unittest.main()
